### ISSUE TITLE
Fun with Floats

### ISSUE THREAD

**Issue by wenkokke** [agda/agda#4892]
*This issue is strictly for documentation purposes. I don't think it's possible to fix this behaviour, whilst keeping floats in Agda. However, it seems like a good idea to at least document the ways you can abuse Float to get non-deterministic behaviour.*

After being compiled with MAlonzo, the following program non-deterministically prints *some* number at runtime, but it's never 15.
```agda
open import Agda.Builtin.IO using (IO)
open import Agda.Builtin.Bool using (Bool; true; false)
open import Agda.Builtin.Equality using (_≡_; refl)
open import Agda.Builtin.Float using (Float)
  renaming ( primFloatDiv to _÷_ 
           ; primFloatTimes to _*_ 
           ; primFloatEquality to _==_ )
open import Agda.Builtin.Nat using (Nat; _+_)
open import Agda.Builtin.String using (String)
  renaming ( primShowNat to show )
open import Agda.Builtin.Unit using (⊤)

postulate putStr : String -> IO ⊤

{-# FOREIGN GHC import qualified Data.Text.IO #-}
{-# COMPILE GHC putStr = Data.Text.IO.putStr #-}

infinity maxFloat : Float
infinity = 1.0 ÷ 0.0
maxFloat = 1.7976931348623157e308

isInfinite : Float → Bool
isInfinite x = x == infinity

unstable : (maxFloat * maxFloat) ÷ maxFloat ≡ infinity
unstable = refl

El : Bool → Set
El false = Bool
El true  = Nat

default : (b : Bool) → El b
default false = false
default true  = 14

inc : El (isInfinite infinity) → Nat
inc x = x + 1

cast : El (isInfinite ((maxFloat * maxFloat) ÷ maxFloat)) 
     → El (isInfinite infinity)
cast x rewrite unstable = x

boom : Nat
boom = inc (cast (default (isInfinite ((maxFloat * maxFloat) ÷ maxFloat))))

main : IO ⊤
main = putStr (show boom)
```
The product `maxFloat * maxFloat` overflows to `Infinity`, so dividing it by `maxFloat` just returns `Infinity`. However, when compiled with optimizations, using GHC, the computation succeeds, probably due to GHC performaing the computation on a register with excess precision. (When compiled with `-O0`,  the program prints 15.)

**Comment by wenkokke** [comment]
I don't think you can get these sorts of problem if you don't allow compile-time computation with `Float`, so I guess we could hide the compile-time evaluation of floating-point primitives behind an `--allow-floats` flag?

**Comment by gallais** [comment]
In which case floats may just as well be postulated rather than built in?

**Comment by nad** [comment]
> In which case floats may just as well be postulated rather than built in?

I guess it might be nice to support some kind of (overloaded) floating-point literals.

I am in favour of moving support for floats from Agda to some library.

**Comment by UlfNorell** [comment]
I think an `--allow-floats` flag is a nice solution. This neatly solves the `--safe` issue as well.

**Comment by wenkokke** [comment]
> In which case floats may just as well be postulated rather than built in?

I'd be really sad if we lost the ability do to floating-point computations at compile-time, 'cuz my current research sorta depends on it.

**Comment by wenkokke** [comment]
We may need a better name than `--allow-floats`, though, since we'll have postulated floats either way, and the flag only controls compile-time evaluation.

**Comment by UlfNorell** [comment]
Yes, for sure. Maybe `--compile-time-floats`?

**Comment by wenkokke** [comment]
I'll have a look at this once #4885 is merged.

**Comment by nad** [comment]
> I think an `--allow-floats` flag is a nice solution. This neatly solves the `--safe` issue as well.

I guess your intention is that `--compile-time-floats` would be incompatible with `--safe`. However, there are still problems with floats. I don't think it is nice that you can get different results by switching from `-O0` to `-O`. (I accept that this change can affect performance, and possibly turn a crashing program into one that completes successfully.)

**Comment by wenkokke** [comment]
@nad: Yup, that's the intention.

I agree that it's not nice, but that's floats for you. The solution to this would be to not compile via GHC, since the problem is inherited from GHC (see [-fexcess-precision](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/using-optimisation.html#ghc-flag--fexcess-precision)). We can try and mitigate it, by forcing `-fno-excess-precision` whenever we can, but that's not an option for the "the 32-bit x86 native code generator".

**Comment by wenkokke** [comment]
Actually, another approach would be to decorate those operations which can cause overflow or underflow with an operation to normalise infinities, either explicitly, or by casting to Word64 and back. We could then have an `--unsafe-floats` operation which disables normalisation of infinities and perhaps NaN normalisation.

I'd probably prefer this?

**Comment by wenkokke** [comment]
I did a little experiment, and truncating to Word64 results in consistent behaviour:
```haskell
import GHC.Float
import Prelude hiding (truncate)

maxValue :: Double
maxValue = 1.7976931348623157e308

truncate :: Double -> Double
truncate = castWord64ToDouble . castDoubleToWord64

overflow :: Double
overflow = (maxValue * maxValue) / maxValue

infinity :: Double
infinity = truncate $ (truncate $ maxValue * maxValue) / maxValue

main :: IO ()
main = do
  print maxValue
  print overflow
  print infinity
```
When compiled with `-O0`, this prints:
```
1.7976931348623157e308
Infinity
Infinity
```
When compiled with `-O1`, this prints:
```
1.7976931348623157e308
1.0000000000000002
Infinity
```

**Comment by guillaumebrunerie** [comment]
> However, there are still problems with floats. I don't think it is nice that you can get different results by switching from `-O0` to `-O`.

Can the issue be reproduced with Haskell/GHC only? (i.e. have an Haskell program that gives different results by switching from `-O0` to `-O`)
If so, is that considered a bug in GHC?

**Comment by wenkokke** [comment]
@guillaumebrunerie Yes, it's a known and documented bug. See [Bugs in GHC](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/bugs.html#bugs-ghc).

**Comment by wenkokke** [comment]
> I did a little experiment, and truncating to Word64 results in consistent behaviour:
> 
> ```haskell
> import GHC.Float
> import Prelude hiding (truncate)
> 
> maxValue :: Double
> maxValue = 1.7976931348623157e308
> 
> truncate :: Double -> Double
> truncate = castWord64ToDouble . castDoubleToWord64
> 
> overflow :: Double
> overflow = (maxValue * maxValue) / maxValue
> 
> infinity :: Double
> infinity = truncate $ (truncate $ maxValue * maxValue) / maxValue
> 
> main :: IO ()
> main = do
>   print maxValue
>   print overflow
>   print infinity
> ```
> 
> When compiled with `-O0`, this prints:
> 
> ```
> 1.7976931348623157e308
> Infinity
> Infinity
> ```
> 
> When compiled with `-O1`, this prints:
> 
> ```
> 1.7976931348623157e308
> 1.0000000000000002
> Infinity
> ```

I'm actually gonna implement this fix in #4885.
